Feature/login final

Zord0m opened the pull request:
Login Implementado: 
Lógica de checagem de e-mail e senha existentes no banco revisadas, o bug do hash estar sendo feito duas vezes, uma quando a conta era criada e outra quando era salva estava afetando a busca no banco, foi resolvido.
Versão antes da mudança pedida de arquitetura e organização, a logica vai ser a mesma porem com melhor organização das pastas e arquivos.

Zord0m committed:
:wrench: feat Nome projeto

alteração de nomes nos documentos iniciais.

Zord0m committed:
:construction: feat: cadastro de usuario

Zord0m committed:
Removendo volumes do Docker e adicionando ao .gitignore

Zord0m committed:
Removendo a pasta coverage e adicionando ao .gitignore

Zord0m committed:
Removendo prettier e padronizando a formatação e identação

Zord0m committed:
tarefa: adicionar package-lock.json ao .gitignore

Zord0m committed:
fix: fixar versões de bibliotecas no package.json

Zord0m committed:
fix : ajuste segundo o PR, troca de mongoose para TypeORM e padronização de formatação.

Zord0m committed:
:sparkles: feat: Login único revisado e implementado, confere credenciais no banco e bug no hash resolvido.

Pedrohsantosg commented:
@samuelcamilocosta

samuelcamilocosta commented on src/config/database.ts:
```diff
@@ -1,22 +1,23 @@
+    type: 'mongodb',
+    url:
+        process.env.DB_CONNECTION_STRING ||
+        'mongodb://reforca_root:j5m966qp7jiypfda@reforca-mongo:27017/admin',
```
Por que declarou as environments diretamente no código? isso não pode não kkk
A intenção da environment é esconder uma chave 'secreta' de conexão do banco.
Dessa forma expôs usuário e senha do banco

samuelcamilocosta commented on src/controller/AuthController.ts:
```diff
@@ -0,0 +1,82 @@
+    resetPassword
+} from '../services/AuthService';
+
+interface CustomError extends Error {
```
É bom já criar uma pasta de interfaces para manter as interfaces lá para poder serem reaproveitadas em outros endpoints do projeto

samuelcamilocosta commented on src/controller/AuthController.ts:
```diff
@@ -0,0 +1,82 @@
+}
+
+// Controlador para login de usuário
+export const login = async (req: Request, res: Response) => {
```
Para o Controller, criar uma classe e exporta-la para acessar as funções te ajudará bastante em manutenções no futuro, pois ajuda até a identificar de qual arquivo veio a função

AuthController.login() sempre lhe dirá mais a respeito do que apenas login(), o nome dado a função pode ser utilizado em várias outras funções do sistema e gerar uma baita confusão posteriormente.

Imagine um sistema grande em que se tenha diversos endpoints de Cadastro ou Exclusão no sistema, essas funções estando presentes em uma classe, sempre te dará um norte de que função ela se refere imediatamente.

samuelcamilocosta commented on src/controller/AuthController.ts:
```diff
@@ -0,0 +1,82 @@
+    const { email, password } = req.body;
+
+    try {
+        const user = await loginUser(email, password); // Faz o login chamando loginUser do AuthService
```
Colocando o loginUser dentro de uma classe chamada AuthService e exportando o AuthService tornará já dispensável o comentário dizendo que esse loginUser se refere ao AuthService.

Então, essa alteração de criar as classes e as funções dentro das classes se faz importante para o AuthService também.

samuelcamilocosta commented on src/models/Users.ts:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    // Método para comparar a senha armazenada com a fornecida
+    async comparePassword(candidatePassword: string): Promise<boolean> {
```
Documente essa função com TSDoc, deixarei um padrão de exemplo aqui, aplique nas outras funções do sistema.
Dito isso, essa função estaria melhor localizada em um repositório

```
/**
 * comparePassword
 *
 * Método para comparar a senha armazenada com a fornecida
 *
 * @params candidatePassword senha que será comparada a senha do usuário
 * 
 * @returns Verdadeiro caso as senhas sejam iguais ou falso caso as senhas sejam diferentes
 */
```

samuelcamilocosta commented on src/models/Users.ts:
```diff
@@ -0,0 +1,61 @@
+
+    // Função para hash da senha antes de inserir no banco
+    @BeforeInsert()
+    async hashPassword() {
```
documentar essa função com TSDoc e tipar o retorno da função.

samuelcamilocosta commented on src/services/AuthService.ts:
Não se esquecer de transformar AuthService em uma classe e criar os repositórios para manter as funções de comunicação com o banco de dados diretamente em uma classe.

samuelcamilocosta replied:
Documentar as funções do arquivo com TSDoc

samuelcamilocosta replied:
Os console.logs podem ser importantes para te ajudar a debuggar o código, porém não há ganho para mantê-los na aplicação, eles apenas irão poluir o seu código e em determinadas situações, ainda pode prejudicar a performance e expor uma informação que deveria ser sigilosa a outros.

No cenário aqui é ainda mais grave, pois você consolou a senha do usuário fornecida, a senha dele criptografada e o e-mail. Ou seja, você terá no terminal os dados de todos os usuários que logaram na aplicação e vazará a senha pessoal que eles tentaram logar na aplicação. 

Isso não pode acontecer de maneira nenhuma, daria acesso a qualquer pessoa que ter acesso ao histórico do terminal, um monte de informações confidenciais.

Dito isso, remova os console.logs do sistema, por favor

samuelcamilocosta commented on src/controller/UserController.ts:
```diff
@@ -0,0 +1,144 @@
+
+export class UserController {
+    // Função para criar um usuário
+    async createUser(req: Request, res: Response) {
```
faltou tipar o retorno da função

Pedrohsantosg committed:
chore(database): mover string de conexão para variável de ambiente

-Removida a string de conexão do código fonte em src/config/database.ts
- Adicionado suporte ao carregamento de variáveis de ambiente usando dotenv
- Atualizado .gitignore para incluir o arquivo .env
- Garantido que a configuração do banco não exponha credenciais sensíveis

Pedrohsantosg committed:
refactor(interfaces): mover CustomError para pasta de interfaces

- Criada pasta src/interfaces para centralizar interfaces reutilizáveis
- Movida a interface CustomError para src/interfaces/CustomError.ts
- Atualizado AuthController.ts para importar a interface do novo local

Pedrohsantosg committed:
refactor(auth-service): migrar funções para classe AuthService

- Refatoradas funções de autenticação para uma classe AuthService
- Atualizado AuthController para utilizar os métodos da nova classe
- Melhorada a organização e manutenibilidade do código

Pedrohsantosg committed:
docs(user-model): adicionar TSDoc completo ao modelo User

- Documentada a classe User e seus métodos com TSDoc
- Melhorada a organização e entendimento do código
- Padronização aplicada para futuras manutenções e desenvolvimento

Pedrohsantosg committed:
refactor(user-model): tipificar função hashPassword com TypeScript

Adicionada tipagem explícita à função hashPassword

Pedrohsantosg committed:
refactor(user-controller): adicionar TSDoc e tipagem completa ao UserController

Documentadas todas as funções do controlador com TSDoc
 Adicionada tipagem explícita aos retornos das funções

Pedrohsantosg committed:
refactor(user-controller): remover console.logs e evitar exposição de dados sensíveis

Removidos todos os console.logs

Pedrohsantosg committed:
feat(errors): adicionar classe CustomError para gerenciamento de erros personalizados

Pedrohsantosg committed:
fix(controllers): substituir any por unknown e melhorar o tratamento de erros

Pedrohsantosg committed:
fix(auth-controller): adicionar função forgotPassword e corrigir rotas

Zord0m committed:
Testando a configuração

samuelcamilocosta commented:
PR aberto no repositório errado
